plugins/parsers/json/parser_test.go

```diff
@@ -1356,3 +1357,47 @@ func TestParseArrayWithWildcardTagKeys(t *testing.T) {
 		})
 	}
 }
+
+// This function with the seeds get run during all unit tests
+// To run this in "fuzzing" mode, where the go fuzzer comes up with all sorts
+// of use cases, run: go test -fuzz=Fuzz ./plugins/parsers/json/...
+func FuzzParserJSON(f *testing.F) {
+	parser := &Parser{MetricName: "testing"}
+	require.NoError(f, parser.Init())
+
+	f.Add([]byte(validJSON))
+	f.Add([]byte(validJSONArray))
+	f.Add([]byte(validJSONArrayMultiple))
+	f.Add([]byte(validJSONArrayTags))
+	f.Add([]byte(validJSONNewline))
+	f.Add([]byte(validJSONTags))
+
+	f.Fuzz(func(t *testing.T, input []byte) {
+		// skip invalid JSON
+		var data interface{}
+		err := json.Unmarshal(input, &data)
+		if err != nil {
+			t.Skip()
+		}
```

Comment:
Since a fuzz test is looking for crashing input, and crashing input doesn't have to be valid json, skipping invalid json might skip cases that would crash.

It could be fine to have one fuzz test that only handles valid json and another that handles any input, but if there's only one test it makes sense to me for it not to skip invalid cases.

Reply:
> skipping invalid json might skip cases that would crash.

To determine if the JSON is valid we use the go standard library JSON parser. One of the comments made in the Gophercon presentation about fuzzing was to not spend our time parsing the standard library.

In order to make the time spent fuzzing useful, it seems wise to get interesting test cases that make their way into calling `parseObject` and `parseArray`

Is there a better way to focus on that code?

Reply:
I agree that it isn't a good use of fuzzing resources to spend time running the standard library. The current code doesn't fuzz test json.Unmarshal but it's similar in that it spends a lot of time validating and throwing out candidate input.

Two ways to do this more efficiently: 1. Don't use `[]byte input` directly as input to parser.Parse, instead use it to generate a valid json syntax tree, render that to text and pass it to parser.Parse. 2. Fuzz test parseObject and parseArray directly. This would also require some generator code because they expect map[string]interface{} and time.Time and the fuzzing framework doesn't provide test values of those types.

The fuzzing book covers generating grammatically valid input in these chapters
https://www.fuzzingbook.org/html/Grammars.html
https://www.fuzzingbook.org/html/GrammarMiner.html

Reply:
Thanks for the call today. The goal and methods are much more clear. Some short notes, please feel free to enhance/correct as necessary:

In general we talked about two high-level types of fuzzing:
* send all data generated by fuzzer directly to parse function
* send only valid data generated using the fuzz input as entropy to generate valid inputs and pass that to parse

When to use each:
1) hand written parsers (e.g. influx) - use both styles
2) library (e.g. json) - prefer valid input

Next steps include a spec with the following for the JSON parser:

* Layout all the various parts and pieces of JSON itself that we would need to generate data for to ensure we are covering all the various parts of the data format. 
* Design method to take in data from the fuzzer, use that data a seed/entropy and generate valid JSON that is then sent into the parser directly
* Worth calling out what the goal of each parsers would be to aid in having others jump in